contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,317 @@
+        _contractGroup[target] = group;
+        emit ContractGroupUpdated(target, group);
+    }
+
+    /**
+     * @dev Sets the target contract to be in the "open" group. All restricted functions in the target contract will
+     * become callable by anyone. The caller must be the default admin.
+     */
+    function setContractOpen(address target) public virtual onlyDefaultAdmin {
+        bytes32 group = _GROUP_OPEN;
+        _contractGroup[target] = group;
+        emit ContractGroupUpdated(target, group);
+    }
+
+    /**
+     * @dev Sets the target contract to be in the "closed" group. All restricted functions in the target contract will
+     * be closed down and disallowed to all. The caller must be the default admin.
+     */
+    function setContractClosed(address target) public virtual onlyDefaultAdmin {
+        bytes32 group = _GROUP_CLOSED;
+        _contractGroup[target] = group;
+        emit ContractGroupUpdated(target, group);
+    }
+
+    /**
+     * @dev Returns true if the team has already been created via {createTeam}.
+     */
+    function _teamExists(uint8 team) internal virtual returns (bool) {
+        return _getTeam(_createdTeams, team);
+    }
+
+    /**
+     * @dev Augmented version of {AccessControl-_grantRole} that keeps track of user team bitmaps.
+     */
+    function _grantRole(bytes32 role, address user) internal virtual override {
+        super._grantRole(role, user);
+        (bool isTeam, uint8 team) = _decodeTeamRole(role);
+        if (isTeam) {
+            require(_teamExists(team));
+            _userTeams[user] = _setTeam(_userTeams[user], team, true);
+        }
+    }
+
+    /**
+     * @dev Augmented version of {AccessControl-_revokeRole} that keeps track of user team bitmaps.
+     */
+    function _revokeRole(bytes32 role, address user) internal virtual override {
+        super._revokeRole(role, user);
+        (bool isTeam, uint8 team) = _decodeTeamRole(role);
+        if (isTeam) {
+            require(_teamExists(team));
+            require(team != _TEAM_ALL);
+            _userTeams[user] = _setTeam(_userTeams[user], team, false);
+        }
+    }
+
+    /**
+     * @dev Returns the {AccessControl} role id that corresponds to a team.
+     *
+     * This role id starts with the ASCII characters `team:`, followed by zeroes, and ends with the single byte
+     * corresponding to the team number.
+     */
+    function _encodeTeamRole(uint8 team) internal virtual returns (bytes32) {
+        return bytes32("team:") | bytes32(uint256(team));
+    }
+
+    /**
+     * @dev Decodes a role id into a team, if it is a role id of the kind returned by {_encodeTeamRole}.
+     */
+    function _decodeTeamRole(bytes32 role) internal virtual returns (bool, uint8) {
+        bytes32 tagMask = hex"ffffffff";
+        bytes32 teamMask = hex"ff";
+        bool isTeam = (role & tagMask) == bytes32("team:");
+        uint8 team = uint8(uint256(role & teamMask));
+        return (isTeam, team);
```

Comment:
```suggestion
        bytes32 tag = (role >> 8) << 8;
        uint8 team  = uint8(role[31]);
        return (tag == bytes32("team:"), team);
```

Reply:
that is not correct though.

- `(role >> 8) << 8` removes the rightmost byte
- `role & ~hex"ff"` removes the leftmost byte

Reply:
You're right, but I meant `role & ~0xff`.

Reply:
`~bytes32(uint256(0xff))` might work.

Reply:
the fact that even we get confused by it says a lot about how messy this is